compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -818,11 +824,48 @@ impl<'tcx> ThirBuildCx<'tcx> {
             })),
 
             hir::ExprKind::OffsetOf(_, _) => {
-                let data = self.typeck_results.offset_of_data();
-                let &(container, ref indices) = data.get(expr.hir_id).unwrap();
-                let fields = tcx.mk_offset_of_from_iter(indices.iter().copied());
+                let offset_of_intrinsic = tcx.require_lang_item(LangItem::OffsetOf, expr.span);
+                let mk_u32_kind = |val: u32| ExprKind::NonHirLiteral {
+                    lit: ScalarInt::try_from_uint(val, Size::from_bits(32)).unwrap(),
+                    user_ty: None,
+                };
+                let mk_call =
+                    |thir: &mut Thir<'tcx>, ty: Ty<'tcx>, variant: VariantIdx, field: FieldIdx| {
+                        let fun_ty =
+                            Ty::new_fn_def(tcx, offset_of_intrinsic, [ty::GenericArg::from(ty)]);
+                        let fun = thir
+                            .exprs
+                            .push(mk_expr(ExprKind::ZstLiteral { user_ty: None }, fun_ty));
+                        let variant =
+                            thir.exprs.push(mk_expr(mk_u32_kind(variant.as_u32()), tcx.types.u32));
+                        let field =
+                            thir.exprs.push(mk_expr(mk_u32_kind(field.as_u32()), tcx.types.u32));
+                        let args = Box::new([variant, field]);
+                        ExprKind::Call {
+                            ty: fun_ty,
+                            fun,
+                            args,
+                            from_hir_call: false,
+                            fn_span: expr.span,
+                        }
+                    };
+
+                let indices = self.typeck_results.offset_of_data().get(expr.hir_id).unwrap();
+                let mut expr = None::<ExprKind<'tcx>>;
+
+                for &(container, variant, field) in indices.iter() {
+                    let next = mk_call(&mut self.thir, container, variant, field);
+                    expr = Some(match expr.take() {
+                        None => next,
+                        Some(last) => {
```

Comment:
ymmv: Consider phrasing this via [`Iterator::reduce`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.reduce), since if it's `indices.iter().map(…).reduce(…).unwrap_or(…)` it handles the "is this the first one" checks for you automatically, which might make it more directly visible what it's doing.

Reply:
Thanks for the suggestion, I did not know that method. However, both closures mutably borrow `self.thir.exprs`, so that's unpractical.